lib/chef/provider/yum_repository.rb

```diff
@@ -46,78 +45,65 @@ def load_current_resource; end
-            block { Chef::Provider::Package::Dnf::PythonHelper.instance.restart }
-          else
-            block { Chef::Provider::Package::Yum::YumCache.instance.reload }
-          end
-          action :nothing
-        end
+        package_cache_reload_resource
       end
 
       action :delete do
         # clean the repo cache first
-        declare_resource(:execute, "yum clean all #{new_resource.repositoryid}") do
+        execute "yum clean all #{new_resource.repositoryid}" do
           command "yum clean all --disablerepo=* --enablerepo=#{new_resource.repositoryid}"
           only_if "yum repolist all | grep -P '^#{new_resource.repositoryid}([ \t]|$)'"
         end
 
-        declare_resource(:file, ::File.join(new_resource.reposdir, "#{new_resource.repositoryid}.repo")) do
+        file ::File.join(new_resource.reposdir, "#{new_resource.repositoryid}.repo") do
           action :delete
-          notifies :create, "ruby_block[package-cache-reload-#{new_resource.repositoryid}]", :immediately
+          notifies :run, "ruby_block[package-cache-reload-#{new_resource.repositoryid}]", :immediately
         end
 
-        declare_resource(:ruby_block, "package-cache-reload-#{new_resource.repositoryid}") do
-          if ( platform?("fedora") && node["platform_version"].to_i >= 22 ) ||
-              ( platform_family?("rhel") && node["platform_version"].to_i >= 8 )
-            block { Chef::Provider::Package::Dnf::PythonHelper.instance.restart }
-          else
-            block { Chef::Provider::Package::Yum::YumCache.instance.reload }
-          end
-          action :nothing
-        end
+        package_cache_reload_resource
       end
 
       action :makecache do
-        declare_resource(:execute, "yum-makecache-#{new_resource.repositoryid}") do
+        execute "yum-makecache-#{new_resource.repositoryid}" do
           command "yum -q -y makecache --disablerepo=* --enablerepo=#{new_resource.repositoryid}"
           action :run
           only_if { new_resource.enabled }
+          notifies :run, "ruby_block[package-cache-reload-#{new_resource.repositoryid}]", :immediately
         end
 
-        declare_resource(:ruby_block, "package-cache-reload-#{new_resource.repositoryid}") do
-          if ( platform?("fedora") && node["platform_version"].to_i >= 22 ) ||
-              ( platform_family?("rhel") && node["platform_version"].to_i >= 8 )
+        package_cache_reload_resource
+      end
+
+      alias_method :action_add, :action_create
+      alias_method :action_remove, :action_delete
+
+      def package_cache_reload_resource
+        ruby_block "package-cache-reload-#{new_resource.repositoryid}" do
+          if ( fedora? && node["platform_version"].to_i >= 22 ) || ( rhel? && node["platform_version"].to_i >= 8 )
             block { Chef::Provider::Package::Dnf::PythonHelper.instance.restart }
           else
             block { Chef::Provider::Package::Yum::YumCache.instance.reload }
           end
```

Comment:
This really shouldn't be a public method, I nearly changed how this behaves awhile back but wasn't quite successful.

`build_resource(:package, "whatever").run_action(:flush_cache)` would be better, which will wind up calling this, but will change whenever the flush_cache action changes.

Reply:
so just flush a dummy package?

Reply:
yeah the name doesn't matter at all, its fake.  its kinda janky, but its the public api we have.

Reply:
oh you'll want to actually declare a package resource with `action :nothing` and then notify it rather than `build_resource`'ing it

Reply:
Updated